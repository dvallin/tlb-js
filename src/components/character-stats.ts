import { ItemKind } from './items'

export type CharacterStatsType = keyof typeof characterStats
export interface CharacterStatsComponent {
  type: CharacterStatsType
  current: CharacterStats
}

export function createCharacterStatsComponent(type: CharacterStatsType): CharacterStatsComponent {
  return { type, current: JSON.parse(JSON.stringify(characterStats[type])) }
}

export type bodyPartType = 'head' | 'arm' | 'torso' | 'leg'

export interface BodyPart {
  type: bodyPartType
  itemAttachments: ItemKind[]
  health: number
}

export interface CharacterStats {
  bodyParts: { [key: string]: BodyPart }
  strength: number
  movement: number
  actions: number
  aim: number
}

function humanoidBodyParts(health: number): { [key: string]: BodyPart } {
  return {
    head: { type: 'head', itemAttachments: ['helmet'], health },
    leftArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health },
    rightArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health },
    torso: { type: 'torso', itemAttachments: ['armor'], health },
    leftLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health },
    rightLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health },
  }
}

export const characterStats = {
  player: {
    bodyParts: humanoidBodyParts(14),
    strength: 5,
    movement: 3,
    actions: 3,
    aim: 7,
  },
  guard: {
    bodyParts: humanoidBodyParts(8),
    strength: 5,
    movement: 3,
    actions: 3,
    aim: 3,
  },
  eliteGuard: {
    bodyParts: humanoidBodyParts(12),
    strength: 8,
    movement: 2,
    actions: 5,
    aim: 6,
  },
}
export const characterStatsTypeguard: { [key: string]: CharacterStats } = characterStats

export function speed(stats: CharacterStatsComponent): number {
  return stats.current.movement / 13
}
