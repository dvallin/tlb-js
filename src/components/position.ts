import { Vector } from '../spatial'

export interface PositionComponent {
  level: number
  position: Vector
}

export function centeredPosition(level: number, position: Vector): PositionComponent {
  return { level, position: position.center }
}
