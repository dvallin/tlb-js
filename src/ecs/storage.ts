
export interface Storage<T> {

    insert(entity: number, value: T): void
    get(entity: number): T | undefined
    remove(entity: number): T | undefined
    has(entity: number): boolean

    foreach(f: (entity: number, value: T) => void): void
    clear(): void
}

export class SetStorage implements Storage<{}> {

    private data: Set<number> = new Set()

    public insert(entity: number, { }: {}): void {
        this.data.add(entity)
    }

    public clear(): void {
        this.data.clear()
    }

    public get(entity: number): {} | undefined {
        return this.data.has(entity) ? {} : undefined
    }

    public remove(entity: number): {} | undefined {
        return this.data.delete(entity) ? {} : undefined
    }

    public has(entity: number): boolean {
        return this.data.has(entity)
    }

    public foreach(f: (entity: number, value: {}) => void): void {
        this.data.forEach(v => f(v, {}))
    }
}

export class MapStorage<T> implements Storage<T> {

    private data: Map<number, T> = new Map()

    public insert(entity: number, value: T): void {
        this.data.set(entity, value)
    }

    public clear(): void {
        this.data.clear()
    }

    public get(entity: number): T | undefined {
        return this.data.get(entity)
    }

    public remove(entity: number): T | undefined {
        const value = this.get(entity)
        this.data.delete(entity)
        return value
    }

    public has(entity: number): boolean {
        return this.data.has(entity)
    }

    public foreach(f: (entity: number, value: T) => void): void {
        this.data.forEach((value, entity) => f(entity, value))
    }
}

export class VectorStorage<T> implements Storage<T> {

    private data: (T | undefined)[] = []

    public insert(entity: number, value: T): void {
        this.data[entity] = value
    }

    public clear(): void {
        this.data = []
    }

    public get(entity: number): T | undefined {
        return this.data[entity]
    }

    public remove(entity: number): T | undefined {
        const value = this.get(entity)
        this.data[entity] = undefined
        return value
    }

    public has(entity: number): boolean {
        return this.data[entity] !== undefined
    }

    public foreach(f: (entity: number, value: T) => void): void {
        for (let entity = 0; entity < this.data.length; entity++) {
            const value = this.data[entity]
            if (value !== undefined) {
                f(entity, value)
            }
        }
    }
}
