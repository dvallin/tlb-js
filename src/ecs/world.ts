import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
