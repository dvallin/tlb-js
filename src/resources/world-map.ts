import { DiscreteSpace, Vector, Space } from '../spatial'
import { Entity } from '../ecs/entity'
import { ResourceName, TlbResource, TlbWorld } from '../tlb'
import { FeatureComponent, features } from '../components/feature'
import { Shape } from '../geometry/shape'
import { SetSpace, DiscreteSetSpace } from '../spatial/set-space'
import { FovComponent } from '../components/fov'
import { Rectangle } from '../geometry/rectangle'

export class WorldMap implements TlbResource {
  public readonly kind: ResourceName = 'map'

  public readonly tiles: Space<Entity> = new DiscreteSpace<Entity>()
  public readonly characters: Space<Entity> = new DiscreteSpace<Entity>()

  public boundaries = new Rectangle(0, 0, 0, 0)

  public visible: SetSpace = new DiscreteSetSpace()
  public readonly discovered: SetSpace = new DiscreteSetSpace()

  public constructor() {}

  public update(world: TlbWorld): void {
    this.visible = new DiscreteSetSpace()
    world.components.get('player')!.foreach(player => {
      const fov = world.getComponent<FovComponent>(player, 'fov')
      if (fov !== undefined) {
        fov.fov.forEach(p => {
          this.visible.set(p)
          this.discovered.set(p)
        })
      }
    })
  }

  public setTile(position: Vector, entity: Entity): void {
    this.tiles.set(position, entity)
    this.boundaries = this.boundaries.cover(position)
  }
  public setCharacter(position: Vector, entity: Entity): void {
    this.characters.set(position, entity)
    this.boundaries = this.boundaries.cover(position)
  }
  public getTile(position: Vector): Entity | undefined {
    return this.tiles.get(position)
  }
  public getCharacter(position: Vector): Entity | undefined {
    return this.characters.get(position)
  }
  public removeTile(position: Vector): Entity | undefined {
    return this.tiles.remove(position)
  }
  public removeCharacter(position: Vector): Entity | undefined {
    return this.characters.remove(position)
  }

  public isTileBlocking(world: TlbWorld, position: Vector): boolean {
    return this.tileMatches(world, position, f => {
      if (f === undefined) {
        return true
      }
      return features[f.type].blocking
    })
  }

  public isDiscovered(position: Vector): boolean {
    return this.discovered.has(position)
  }

  public isLightBlocking(world: TlbWorld, position: Vector): boolean {
    return this.tileMatches(world, position, f => {
      if (f === undefined) {
        return true
      }
      return features[f.type].lightBlocking
    })
  }

  public tileMatches(world: TlbWorld, position: Vector, predicate: (f: FeatureComponent | undefined) => boolean): boolean {
    const entity = this.tiles.get(position)
    let feature: FeatureComponent | undefined
    if (entity !== undefined) {
      feature = world.getComponent<FeatureComponent>(entity, 'feature')
    }
    return predicate(feature)
  }

  public isShapeFree(world: TlbWorld, shape: Shape): boolean {
    return this.shapeHasAll(world, shape, f => f === undefined)
  }

  public isShapeBlocked(world: TlbWorld, shape: Shape): boolean {
    return this.shapeHasSome(world, shape, f => f !== undefined)
  }

  public shapeHasAll(world: TlbWorld, shape: Shape, predicate: (f: FeatureComponent | undefined) => boolean): boolean {
    return shape.all(p => this.tileMatches(world, p, predicate))
  }

  public shapeHasSome(world: TlbWorld, shape: Shape, predicate: (f: FeatureComponent | undefined) => boolean): boolean {
    return shape.some(p => this.tileMatches(world, p, predicate))
  }
}
