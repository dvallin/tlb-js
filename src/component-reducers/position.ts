import { WorldMapResource } from '../resources/world-map'
import { Vector } from '../spatial'
import { TlbWorld } from '../tlb'
import { Entity } from '../ecs/entity'
import { PositionComponent } from '../components/position'

export function move(world: TlbWorld, entity: Entity, level: number, position: Vector): void {
  const l = world.getResource<WorldMapResource>('map').levels[level]
  if (l !== undefined && !l.isBlocking(world, position, entity)) {
    const oldPosition = world.getComponent<PositionComponent>(entity, 'position')!
    l.removeCharacter(oldPosition.position)
    l.setCharacter(position, entity)
    world.editEntity(entity).withComponent<PositionComponent>('position', { level, position })
  }
}
