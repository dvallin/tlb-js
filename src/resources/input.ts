import { TlbResource, ResourceName } from '../tlb'
import { Vector } from '../spatial'
import { Position } from '../renderer/position'
import { KEYS } from 'rot-js'

export interface Input {
  position: Position | undefined
  mouseDown: boolean
  mousePressed: boolean
  mouseReleased: boolean

  keyDown: Set<number>
  keyPressed: Set<number>
  keyReleased: Set<number>

  createMovementDelta(): Vector
}

export class InputResource implements TlbResource, Input {
  public readonly kind: ResourceName = 'input'

  public position: Position | undefined = undefined

  public mouseDown: boolean = false
  public mousePressed: boolean = false
  public mouseReleased: boolean = false

  public keyDown: Set<number> = new Set()
  public keyPressed: Set<number> = new Set()
  public keyReleased: Set<number> = new Set()

  private mouseEvent: MouseEvent | undefined = undefined
  private keyEvents: KeyboardEvent[] = []

  public constructor(public readonly eventToPosition: (event: MouseEvent | TouchEvent) => Position | undefined) {
    document.addEventListener('mousedown', e => (this.mouseEvent = e))
    document.addEventListener('mousemove', e => (this.mouseEvent = e))
    document.addEventListener('keydown', e => this.keyEvents.push(e))
    document.addEventListener('keyup', e => this.keyEvents.push(e))
  }

  public update(): void {
    this.handleMouseEvent()
    this.handleKeyboardEvents()
  }

  public createMovementDelta(): Vector {
    let delta = new Vector(0, 0)
    if (this.keyDown.has(KEYS.VK_H)) {
      delta = delta.add(Vector.fromDirection('left'))
    }
    if (this.keyDown.has(KEYS.VK_J)) {
      delta = delta.add(Vector.fromDirection('down'))
    }
    if (this.keyDown.has(KEYS.VK_K)) {
      delta = delta.add(Vector.fromDirection('up'))
    }
    if (this.keyDown.has(KEYS.VK_L)) {
      delta = delta.add(Vector.fromDirection('right'))
    }
    return delta.normalize()
  }

  private handleMouseEvent(): void {
    this.mouseReleased = false
    this.mousePressed = false
    if (this.mouseEvent) {
      this.position = this.eventToPosition(this.mouseEvent)
      const pressed = this.mouseEvent.buttons > 0
      if (pressed) {
        if (this.position) {
          this.mousePressed = true
          this.mouseDown = true
        }
      } else {
        this.mouseDown = false
        this.mouseReleased = true
      }
      this.mouseEvent = undefined
    }
  }

  private handleKeyboardEvents(): void {
    this.keyPressed.clear()
    this.keyReleased.clear()
    for (const e of this.keyEvents) {
      if (e.type === 'keydown') {
        this.keyPressed.add(e.keyCode)
        this.keyDown.add(e.keyCode)
      } else {
        this.keyReleased.add(e.keyCode)
        this.keyDown.delete(e.keyCode)
      }
    }
    this.keyEvents = []
  }
}
