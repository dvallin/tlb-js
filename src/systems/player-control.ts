import { ComponentName, TlbSystem, ResourceName, SystemName } from '../tlb'
import { World } from '../ecs/world'
import { PositionComponent } from '../components/position'
import { Input } from '../resources/input'
import { WorldMap } from 'src/resources/world-map'

export class PlayerControl implements TlbSystem {
  public readonly components: ComponentName[] = ['player', 'position']

  public update(world: World<ComponentName, SystemName, ResourceName>, entity: number): void {
    const position = world.getComponent<PositionComponent>(entity, 'position')!
    const input = world.getResource<Input>('input')
    const delta = input.createMovementDelta()
    if (delta.squaredLength() > 0) {
      const newPosition = position.position.add(delta.mult(0.13))
      const map = world.getResource<WorldMap>('map')
      if (!map.isTileBlocking(world, newPosition.floor())) {
        map.characters.remove(position.position.floor())
        map.characters.set(newPosition.floor(), entity)
        world.editEntity(entity).withComponent('position', { position: newPosition })
      }
    }
  }
}
