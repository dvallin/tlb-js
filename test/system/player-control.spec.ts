import { PlayerControl } from '../../src/systems/player-control'
import { TlbWorld } from '../../src/tlb'
import { World } from '../../src/ecs/world'
import { Input } from '../../src/resources/input'
import { VectorStorage } from '../../src/ecs/storage'
import { Vector } from '../../src/spatial'
import { PositionComponent } from '../../src/components/position'
import { WorldMapResource } from '../../src/resources/world-map'
import { mockMap, mockReturnValue, mockInput } from '../mocks'
import { CharacterStatsComponent, createCharacterStatsComponent } from '../../src/components/character-stats'

describe('PlayerControl', () => {
  let system: PlayerControl
  let world: TlbWorld
  let input: Input
  let map: WorldMapResource
  beforeEach(() => {
    system = new PlayerControl()
    world = new World()
    input = mockInput(world)
    map = mockMap(world)
    world.registerComponentStorage('position', new VectorStorage<PositionComponent>())
    world.registerComponentStorage('character-stats', new VectorStorage<CharacterStatsComponent>())
    world
      .createEntity()
      .withComponent('position', { position: new Vector([2, 3]) })
      .withComponent<CharacterStatsComponent>('character-stats', createCharacterStatsComponent('player'))
  })

  it('only adds delta if non blocking', () => {
    input.createMovementDelta = jest.fn().mockReturnValue(new Vector([-1, 0]))
    mockReturnValue(map.isBlocking, true)

    system.update(world, 0)

    expect(world.getComponent(0, 'position')).toEqual({ position: new Vector([2, 3]) })
  })

  describe('movement allowed', () => {
    beforeEach(() => {
      input.createMovementDelta = jest.fn().mockReturnValue(new Vector([-1, 0]))
      mockReturnValue(map.isBlocking, false)

      system.update(world, 0)
    })

    it('adds delta times speed', () => {
      expect(world.getComponent(0, 'position')).toEqual({ position: new Vector([1.7692307692307692, 3]) })
    })

    it('moves character', () => {
      expect(map.removeCharacter).toHaveBeenCalledWith(new Vector([2, 3]))
      expect(map.setCharacter).toHaveBeenCalledWith(new Vector([1, 3]), 0)
    })
  })
})
