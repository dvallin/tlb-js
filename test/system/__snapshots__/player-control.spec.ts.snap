// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerControl moves player position 1`] = `
Vector {
  "coordinates": Array [
    0.8076923076923077,
    0.25,
  ],
}
`;
